### ISSUE TITLE
Speech synthesis splits words at consecutive SPAN tags boundaries

### ISSUE THREAD

**Issue by sjabsr** [edrlab/thorium-reader#2038]
We are a Swiss-French non-profit lending audio book and epubs to reading impaired people. We explore a new format that is in demand in the (French) dyslexic community: syllabic colorization. We did a proof of concept EPUB. The colorization is done with in-line or linked CSS, in either case with SPAN tags. E.g. : "croissant" is colorized as : 
`<span style="color:red">crois</span><span style="color:blue">san</span><span style="color:gray">t</span>`

We've tested it in Thorium : the text rendering was perfect (unlike in Easy Reader's, in which consecutive SPAN tags are rendered with a space between them), but Thorium's speech synthesis seems to treat each SPAN as a separate word, disrupting the flow and pronunciation of the language, thus rendering the audio output unintelligible for our intended users.

I imagine this use case is pretty niche right now, but it could grow in the future.

**Comment by gautierchomel** [comment]
Syllabic colorization should be done at Reading System level, unfortunately we've not yet secured funding for this.

**Comment by sjabsr** [comment]
I concur, and great to know it's foreseen the right way !

**Comment by sjabsr** [comment]
In my enthusiasm for those features I forgot the subject of the issue reported here : as a SPAN within a word shouldn't be treated as a word boundary, isn't this still an issue with Reader's TTS ? A creative publication might use text colors to convey some meaning or a creative aspect to an epub, and currently such designs will "break" the TTS.

**Comment by gautierchomel** [comment]
Yes, it's a known limitation of TTS related to the fact that any tagging is considered a separate portion and sent apart to the voice API. This is necessary for example when the portion of text is affected by a different language tagging (so the TTS can switch to the adequate voice).

**Comment by danielweck** [comment]
this should work correctly in Thorium, as we implement an utterance-building algorithm that treats inline span elements as non-breaking, unless they carry a different language attribute to the enclosing or ancestor element context.

would you mind sharing the epub so I can investigate? thank you very much.

feel free to use my email address:
daniel.weck@gmail.com

**Comment by danielweck** [comment]
closing due to inactivity / lack of testing material

**Comment by danielweck** [comment]
Thank you for privately sharing `Confessions d'une dyslexique-Colorisation syllabique.epub` :)
I tested it with Thorium's TTS readaloud feature (latest version `3.1.0`, everything seems to work fine).
See screenshots (with and without "sentence detection / splitting"):

<img width="817" alt="Image" src="https://github.com/user-attachments/assets/94b6446a-d44b-48a1-94a9-a336bcf2ae57" />
<img width="817" alt="Image" src="https://github.com/user-attachments/assets/ed399815-429e-4be0-925a-1e29a8455647" />

**Comment by sjabsr** [comment]
Thank you for looking into it ! I does indeed work. 
Sorry, I should have checked the current situation before to re-send my quite old email !

**Comment by danielweck** [comment]
that's absolutely fine, thank you for your feedback! :)

**Comment by sjabsr** [comment]
If I may ask here, in the context of this colorized text, where the color of the text matters the most as it is being read : is there any way to override the style of the TTS highlighting style, so as to keep the colorization while the sentence is being read by the TTS ? 
I've tested with the MO vocabulary (media:active-class+CSS), but it doesn't seem to work.

**Comment by danielweck** [comment]
The EPUB3 Media Overlays CSS class system isn't compatible with TTS readaloud ... in MO, the active playback text fragment is always an HTML Element referenced by ID, which can be styled in CSS via the class reference. Typically the Element gets colorized (background, sometimes foreground) but content creators can apply more "exotic" styles such as font weight, geometric transformations, animations, etc. :)

The TTS readaloud highlighting style currently cannot be customised. The following could potentially be configured: solid yellow background (choice of background and foreground colours), orange underline (could be turned off completely, or just use the same colour as the solid background). The current colour scheme was created to work relatively well in all of Thorium's colour themes (dark, sepia, etc.).

 Fun fact: the highlighting engine is the same used for the annotations feature, so in principle other types could be used instead of solid and underline: outline and strikethrough. 

Other interesting fact: a previous incarnation of the highlights engine was applying a CSS mix-blend-mode to the solid background (essentially a DIV overlay on top of the text), but this has severe limitations notably in terms of contrast ratio with foreground/text colour, which would have been evident with your multi-coloured text demo (DYS parts of speech emphasis, or de-emphasis with faded text). The revised highlights engine allows us to control the contrast ratio between text and background, but we have lost the ability to visually show overlapping highlights by blending them (some sort of opacity layering). Tradeoffs...

**Comment by sjabsr** [comment]
Thanks for the info and context ! I the text colorization case, the original material asks for a certain type of highlight styling, so a reader setting would not be optimal (ie we'd need to explain colorized text users how to change the setting). 

I guess  that "CSS customisable TTS highlighting" is not something defined in the epub standard (yet), otherwise Thorium would have implemented it ?

**Comment by HadrienGardeur** [comment]
A reading ruler decoration for the current utterance would work best for this kind of example IMO.

With improvements to Readium Speech for Thorium Web and plans for a brand new Thorium Mobile app on iOS this year, we plan on tackling these issues. As @danielweck pointed out, they're very much open ended and we plan on having an approach that gives control to the user over all of these things.

> Thanks for the info and context ! I the text colorization case, the original material asks for a certain type of highlight styling, so a reader setting would not be optimal (ie we'd need to explain colorized text users how to change the setting).
>
> I guess that "CSS customisable TTS highlighting" is not something defined in the epub standard (yet), otherwise Thorium would have implemented it ?

There's no such thing in EPUB and authored styles are never perfect either. They might work for some users, but it's important to make sure that users can always select something that works best for them.

FYI, we're also looking into syllabic colorization as a feature that could be applied by the reading system, rather than authored.

**Comment by sjabsr** [comment]
> A reading ruler decoration for the current utterance would work best for this kind of example IMO.

A reading ruler would indeed be neat for colorized text ! 

> There's no such thing in EPUB and authored styles are never perfect either. They might work for some users, but it's important to make sure that users can always select something that works best for them.

Yes, I agree that the precedence, for styling choices in this case, of "User > Author > App" is the best of both worlds for control and adaptability.

> FYI, we're also looking into syllabic colorization as a feature that could be applied by the reading system, rather than authored.

I've heard about it, yes, it would be an incredible feature, and I'm impressed that it's planned, as the tools I've seen have quite complex settings panels, due to the different colorization types+goals, e.g. : 

- MS Office toolbar [colorization.ch](url)
- LibreOffice toolbar / web [lirecouleur.arkaline.fr/utilisation](url)

And these would be multiplied by the number of supported languages most probably.

**Comment by danielweck** [comment]
> "User > Author > App" is the best of both worlds for control and adaptability.

Always :)
👍 

Synonym we use whenever we discuss this class of problems: "User > Publisher > Reading System"

**Comment by danielweck** [comment]
On MacOS with Books.app the dark mode removes publisher styles (understandable) but their rendering engine leaves some weird blue word segments at the beginning of sentences.

<img width="709" alt="Image" src="https://github.com/user-attachments/assets/eb03626c-5ba3-46d9-b0b3-c26d5d0f236a" />

<img width="709" alt="Image" src="https://github.com/user-attachments/assets/27afcd09-d833-4356-91df-95c62d4240dc" />

**Comment by danielweck** [comment]
@sjabsr I am posting the screenshots here for future visitors of Thorium's issue tracker :)

Source: https://github.com/edrlab/thorium-reader/discussions/2817

<img width="1840" alt="Screenshot 2025-02-23 at 10 32 44" src="https://github.com/user-attachments/assets/933b46da-a9d0-4fab-8ab7-0ac4dc4a6487" />

<img width="1840" alt="Screenshot 2025-02-23 at 10 33 15" src="https://github.com/user-attachments/assets/a7bbe195-b85c-42c5-857a-d734113e5056" />

<img width="1840" alt="Screenshot 2025-02-23 at 10 33 51" src="https://github.com/user-attachments/assets/b2238599-f136-442e-937d-fa2117e96781" />

<img width="1840" alt="Screenshot 2025-02-23 at 10 34 36" src="https://github.com/user-attachments/assets/e2f34ac8-90a0-4fdf-a605-f87cf2975b0e" />

**Comment by sjabsr** [comment]
Sorry for the late reaction, but beautiful work ! 
And thanks for the heads up in the other issue #2817 !
